gbpservice/neutron/nsf/core/main.py

```diff
@@ -0,0 +1,705 @@
+import os
+import time
+import sys
+import copy
+import threading
+
+import eventlet
+eventlet.monkey_patch()
+
+from multiprocessing.queues import Queue
+from Queue import Empty, Full
+from multiprocessing import Process, Queue, Lock
+import multiprocessing as multiprocessing
+
+from neutron.agent.common import config
+from neutron.common import config as common_config
+from neutron import context
+from neutron.agent import rpc as agent_rpc
+
+from gbpservice.neutron.nsf.core import cfg as core_cfg
+from gbpservice.neutron.nsf.core import lb as core_lb
+from gbpservice.neutron.nsf.core import threadpool as core_tp
+from gbpservice.neutron.nsf.core import periodic_task as core_periodic_task
+from gbpservice.neutron.nsf.core import queue as core_queue
+
+from neutron.common import rpc as n_rpc
+
+from oslo_log import log as logging
+from oslo_config import cfg
+from oslo_service import periodic_task as oslo_periodic_task
+from oslo_service import loopingcall
+from oslo_service import service as os_service
+
+LOG = logging.getLogger(__name__)
+
+process = os.getpid()
+
+
+def is_class(obj):
+    return 'class' in str(type(obj))
+
+
+def name(obj):
+    if callable(obj):
+        return "{0}.{1}.{2}".format(
+            type(obj.im_self).__module__,
+            type(obj.im_self).__name__,
+            obj.__name__)
+    elif is_class(obj):
+        return "{0}.{1}".format(
+            type(obj).__module__,
+            type(obj).__name__)
+    else:
+        return obj.__name__
+
+
+def identify(obj):
+    try:
+        return "(%s)" % (name(obj))
+    except:
+        return ""
+
+
+class AgentReportState(object):
+
+    def __init__(self, agent_state):
+        self._context = context.get_admin_context_without_session()
+        self._agent_state = agent_state
+        self._report_topic = agent_state['plugin_topic']
+        self._report_interval = agent_state['report_interval']
+        self._state_rpc = agent_rpc.PluginReportStateAPI(
+            self._report_topic)
+
+    def report_state(self):
+        try:
+            LOG.debug(_("Reporting state ", self._agent_state))
+            self._state_rpc.report_state(self._context, self._agent_state)
+            self._agent_state.pop('start_flag', None)
+        except AttributeError:
+            # This means the server does not support report_state
+            LOG.warn(_("Neutron server does not support state report."
+                       " Agent State reporting will be "
+                       "disabled."))
+            self._heartbeat.stop()
+            return
+        except Exception:
+            LOG.exception(_("Stopped reporting agent state!"))
+
+
+class RpcAgent(n_rpc.Service):
```

Comment:
Yes, this code is required for neutron agent shows up in the neutron agent-list.
The way you are suggesting is difficult to implement.
There is only a single module, but not single agent, it will register multiple agents and multiple RpcEndpoints for better handling. And each agent like FW, VPN, LB has different data to report to the plugin. Please look at "config_agent_modules" git branch and see modules/config_agent.py